### ISSUE TITLE
n-1 Bug: PCM Analysis does not follow the original specification

### ISSUE THREAD

**Issue by sebinside** [DataFlowAnalysis/DataFlowAnalysis#78]
The current version of the PCM-based analysis does not follow to original specification from [here](https://publikationen.bibliothek.kit.edu/1000151563). Violations of confidentiality requirements are found one data flow diagram node, i.e., one entry in the action sequence to early, i.e., in node `n - 1`. This becomes visible when comparing the PCM-based analysis with the vanilla DFD analysis.

The original specification states: "The label comparison function compares the classification label of incoming data with the clearance label of the receiving node." (p. 97). This can happen because we do not generate an exit node for SEFFs, so the check happens "implicitly" on the out pin of the node `n - 1`.

TL;DR, we propose two changes:
* Add the generation of SEFF exit nodes, see p. 174
* The constraint should be evaluated on the state of the input pin, not the output pin

**Comment by Entenwilli** [comment]
There are two approaches to fixing this bug:
1. Save both characteristics at the input pin and the output pin of the node, and return characteristics at input pin as default. This would also allow access/constraints on output characteristics of a node
2. Save only incoming characteristics, evaluating a node returns both the new node and the propagated outgoing characteristics

Personally, I would prefer the first approach, as it would allow us to express more powerful constraints, simplify the evaluation of nodes and have a representation closer to nodes in a DFD. What do you think @sebinside @Nicolas-Boltz?

**Comment by sebinside** [comment]
I discussed this with Nicolas and we also like the first solution as it is very transparent and shows what actually happens with more detail. However, I won't enable defining constraints that rely on output characteristics for now as this just confuses the user. But having the data for debugging or future analysis extensions is probably a good idea.
